PIPS/pip-1.md

```diff
@@ -0,0 +1,133 @@
+
+```
+constructor (address lockProxyContractAddress, uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash) public ERC20Detailed("ONT Token", "ONTX", 0) {
+    uint256 totalSupply = 1000000000;
+    _mint(lockProxyContractAddress, totalSupply);
+    lockProxy.delegateAsset(nativeChainId, nativeLockProxy, nativeAssetHash, totalSupply);
+}
+```
+
+In the LockProxy, `delegateAsset` can be implemented as:
+```
+struct DelegatedAsset {
+    uint64 nativeChainId;
+    bytes nativeLockProxy;
+    bytes nativeAssetHash;
+}
+
+mapping(address => DelegatedAsset) delegatedAssets;
+mapping(bytes32 => uint256) balances;
+
+function delegateAsset(uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash, uint256 totalSupply) public {
+    address assetHash = _msgSender();
+
+    require(nativeChainId != 0);
+    require(delegatedAssets[assetHash].nativeChainId == 0);
+    require(getBalanceFor(assetHash) == totalSupply);
+
+    // `hash` is any supported hashing function of the respective blockchain
+    bytes32 key = hash(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+
+    require(balances[key] == 0);
+    balances[key] = totalSupply;
+
+    delegatedAssets[_msgSender()] = DelegatedAsset(nativeChainId, nativeLockProxy, nativeAssetHash);
+}
+```
+
+This mapping will be used in the `unlock` function to ensure that delegated tokens will be unlocked only if the correct corresponding token is locked.
+
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed.
+
+The `lock` function can be modified to require the following parameters:
+- address fromAssetHash
+- uint64 toChainId
+- bytes memory targetProxyHash
+- bytes memory toAssetHash
+- bytes memory toAddress
+- uint256 amount
+
+The LockProxy contract should maintain a `balances` mapping of (bytes32 => uint256).
+This `balances` mapping should be updated within the `lock` function:
+```
+// help prevent user mistakes by ensuring delegated assets only get transferred
+// to the native chainId, LockProxy and assetHash that was originally specified
+DelegatedAsset dAsset = delegatedAssets[fromAssetHash];
+if (dAsset.nativeChainId != 0) {
+    require(
+        dAsset.nativeChainId == toChainId &&
+        dAsset.nativeLockProxy == targetProxyHash &&
+        dAsset.nativeAssetHash == toAssetHash
+    );
+}
+
+bytes32 key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
+
+balances[key] += amount;
+
+address fromContractAddr = address(this);
+TxArgs memory txArgs = TxArgs({
+    fromContractAddr: fromContractAddr,
+    fromAssetHash: fromAssetHash,
+    toAssetHash: toAssetHash,
+    toAddress: toAddress,
+    amount: amount
+});
+bytes memory txData = _serializeTxArgs(txArgs);
+
+emit LockEvent(fromContractAddr, fromAssetHash, toChainId, toAssetHash, toAddress, amount);
```

Comment:
Suggest the event follows the format of LockEvent(address fromAssetHash, address fromAddress, uint64 toChainId, bytes toAssetHash, bytes toAddress, uint256 amount, ...optional). The reason behind that is polynetwork explorer will support display of the cross chain asset transaction details, and the ahead six fields need to be fromAssetHash, fromAddress, toChainId, toAsssetHash, toAddress, and amount to make the event be recognized by the explorer. As for fromContractAddr or any other info, can be put at the optional position.

Reply:
sure, will change it